### PR TITLE
Documentation for the testnode role.

### PR DESCRIPTION
**Pull request by andrewschoen**

### CODE REVIEW

File: roles/testnode/README.rst [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+This host is used by teuthology to download ceph packages and will be given higher priority on apt systems::
+
+    gitbuilder_host: gitbuilder.ceph.com
+
+The mirror to download and install ``pip`` from::
+
+    pip_mirror_url: "http://{{ mirror_host }}/pypi/simple"
+
+As hash defining yum repos that would be common across a major version. Each key in the hash will be
```
Comment (zmc) [inline roles/testnode/README.rst]: *A hash

**Comment by zmc** [comment]
This is fantastic!

File: roles/testnode/README.rst [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+- Ubuntu Trusty
+- Ubuntu Vivid
+
+**NOTE:** This role was first created as a port of ceph-qa-chef_.
+
+Usage
++++++
+
+The testnode role is primarily used by the ``testnodes.yml`` playbook.  This playbook is run by cobbler during
+bare metal imaging to prepare a node for testing and it also used by Teuthology during test runs to ensure the config
```
Comment (dmick) [inline roles/testnode/README.rst]: bare-metal

File: roles/testnode/README.rst [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+is correct before testing.
+
+**NOTE:** ``testnodes.yml`` will only run against nodes in the ``testnodes`` group.
```
Comment (dmick) [inline roles/testnode/README.rst]: Change to "is limited to run against hosts in the ``testnodes`` group by a hosts: key in the playbook (free opportunity to reinforce how that works)

File: roles/testnode/README.rst [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+Variables
++++++++++
+
+Available variables are listed below, along with default values (see ``defaults/main.yml``). This role also
```
Comment (dmick) [inline roles/testnode/README.rst]: "(see each role's defaults/main.yml)"

**Comment by andrewschoen** [comment]
Ok, I think I've got all the feedback incorporated here.  I also removed the couple mentions of tags from the ``common`` role.  The ``testnodes.yml`` playbook is the best way to use those tags from the ``common`` role, but I feel those are better documented by that role's README.

File: roles/testnode/README.rst [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+    git_mirror_host: git.ceph.com
+
+This host is used by teuthology to download ceph packages and will be given higher priority on apt systems::
+
+    gitbuilder_host: gitbuilder.ceph.com
+
+The mirror to download and install ``pip`` from::
+
+    pip_mirror_url: "http://{{ mirror_host }}/pypi/simple"
```
Comment (dmick) [inline roles/testnode/README.rst]: Will this work if mirror_host is undefined (i.e. "don't mirror it")?  Should it?

Comment (andrewschoen) [inline roles/testnode/README.rst]: Currently this must be defined.  I'd like mirror_host, gitbuilder_host and
pip_mirror_url to all noop eventually if they aren't provided. Then we'd
change all the defaults here to empty strings.

File: roles/testnode/README.rst [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+A hash defining yum repos that would be common across a major version. Each key in the hash will be
+the name of a yum repo created in /etc/yum.repos.d. The key/value pairs as the value for that repo
```
Comment (dmick) [inline roles/testnode/README.rst]: "name" is a tricky term here.  There's the name of the file, and then the file may contain multiple repos, actually, each with their own [] key, and a separate name= attribute, so there's three names; we should make clear which is which here (and I see that name: key in the actual text).  Documentation is such fun.

File: roles/testnode/README.rst [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    # An example: 
+    common_yum_repos:
+      rhel-7-fcgi-ceph:
+        name: "RHEL 7 Local fastcgi Repo"
+        baseurl: http://gitbuilder.ceph.com/mod_fastcgi-rpm-rhel7-x86_64-basic/ref/master/
+        enabled: 1
+        gpgcheck: 0
+        priority: 2
+
+A hash defining version specific yum_repos. Each key in the hash will be
```
Comment (dmick) [inline roles/testnode/README.rst]: version-specific

File: roles/testnode/README.rst [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    # An example:
+    yum_repos:
+      fedora-fcgi-ceph:
+        name: Fedora Local fastcgi Repo
+        baseurl: http://gitbuilder.ceph.com/mod_fastcgi-rpm-fedora20-x86_64-basic/ref/master/
+        enabled: 1
+        gpgcheck: 0
+        priority: 0
+
+A list of packages to install that is specific to a distro version.  These lists are defined in the var files in ``vars/``::
```
Comment (dmick) [inline roles/testnode/README.rst]: I think maybe a specific example of a particular package that's "specific to a distro version" and one that's "common to a distro or distro version" would be helpful.

File: roles/testnode/README.rst [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+A list of packages to remove. These lists are defined in the var files in ``vars/``::
+
+    packages_to_remove: []
+
+A list of packages to upgrade. These lists are defined in the vars files in ``vars/``::
+
+    packages_to_upgrade: []
+
+The user that teuthology will use to connect to testnodes. This user will be created by this role and assigned to the appropriate groups.
+Eventhough this variable exists, teuthology is not quite ready to support a configurable user::
```
Comment (dmick) [inline roles/testnode/README.rst]: Even though

File: roles/testnode/README.rst [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+    teuthology_user: "ubuntu"
+
+This user is created for use in running xfstests from ceph-qa-suite::
+
+    xfstests_user: "fsgqa"
+
+This will control whether or not rpcbind is started before nfs.  Some distros require this, others don't::
```
Comment (dmick) [inline roles/testnode/README.rst]: Is this based on whether rpcbind is already running?...in which case, can it be automatic?

File: roles/testnode/README.rst [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+    start_rpcbind: true
+
+Set to true if /etc/fstab must be modified to persist things like mount options, which is useful for long lived
```
Comment (dmick) [inline roles/testnode/README.rst]: long-lived. and I still think there's no reason this couldn't happen on every machine, but...that's another bug

File: roles/testnode/README.rst [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+
+A list of ntp servers to use::
+
+    ntp_servers:
+      - 0.us.pool.ntp.org
+      - 1.us.pool.ntp.org
+      - 2.us.pool.ntp.org
+      - 3.us.pool.ntp.org
+
+The lab domain to strip from hostname on rhel 7.x nodes. This is only done if ``lab_domain``
```
Comment (dmick) [inline roles/testnode/README.rst]: why would you want/need to strip a domain from a hostname, and why only for rhel 7.x?

File: roles/testnode/README.rst [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+user
+    Manages the ``teuthology_user`` and ``xfstests_user``. 
+
+Dependencies
+++++++++++++
+
+This role depends on the following roles:
+
+secrets
+    Provides a var, ``secrets_path``, containing the path to where our secrets are located.
```
Comment (dmick) [inline roles/testnode/README.rst]: "to where our secrets are located" -> "of the secrets repository, a tree of ansible variable files"

File: roles/testnode/README.rst (hunk 5 above)
Comment (ktdreyer) [inline roles/testnode/README.rst]: That or, you know, just move all of this stuff into `defaults.yml` :D

File: roles/testnode/README.rst [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+    
+sudo
+    Sets ``ansible_sudo: true`` for this role.
```
Comment (dmick) [inline roles/testnode/README.rst]: ...which causes all the plays "in" (is that the right term) this role to execute with sudo

File: roles/testnode/README.rst (hunk 10 above)
Comment (andrewschoen) [inline roles/testnode/README.rst]: I can't remember the exact circumstance for this, but I'm sure we can make this smarter eventually.

File: roles/testnode/README.rst (hunk 12 above)
Comment (andrewschoen) [inline roles/testnode/README.rst]: I wish I knew, I suspect it had something to do with how we used to cobbler rhel nodes but don't know for certain.  This is one of the ceph-qa-chef quirks we need to revisit eventually.

**Pull request merged**